### DISCUSSION TITLE
How-To test changes to an external library in your solution

### DISCUSSION THREAD

**Discussion by Brennan1994** [HydrologicEngineeringCenter/HEC-FDA#325] [Show and tell]
This article is born of an issue we had where a bug in Framework, an external dependency to FDA, had a bug that was causing strange behavior inside FDA. This discussion is to serve as an example of how we went about fixing that issue, testing it locally, pushing changes, and incorporating it all back into main. 

1. Pull the source code for the external dependency. In this case, that was Framework, available under the HEC Organization on Github. 
2.  Build the solution for the external. Make sure it works as expected. In this case, we had no issues. Solutions should build right out of the box if set up properly. 
3. Commit all changes in your primary solution. Just so you always have the power to undo-all and go back if you mess something up.

4. Right click your solution file in the solution explorer of your main project, and select Add | Existing Project. Navigate to the .csproj file of the external projects you'll want to be changing. Adding an existing project doesn't move the folder,(and so doesn't add those files to the git repository of your main project) but adds the path to the project in your solution file. 
5. Now, remove your references to the NuGet Package repo (in our case, we removed the references for MVVMFramework.Base, and MVVMFramework.ViewModel) and replace them with project references to the projects you just added to your solution. 
6. At this point you should rebuild your solution and ensure that everything works the same as it did before. All you've done is swap a reference that used to look to a nuget repository, to the exact same reference built on your own desktop. 

7. Now, you can interact with both sets of source code in the same IDE Instance and debug in as necessary. When you want to make a change, you can do it, and rebuild and have it immediately integrated. When you're satisfied with your changes, commit them and push to a new branch in the external repository. Make sure not to commit the changes to your proj files and solution file in your main project however. Those changes are the changes in dependencies for testing, and you'll want to undo those when you're finished. 

8. When you're finished, create a pull request for the external library, and merge into main. Github should automatically test and publish a new package you can reference by NuGet moving forward.  Reference that package in your main project before creating your next pull request.

If you have a better method, or questions, please add to this discussion.

**Comment by rnugent3** [comment]
Thanks, @Brennan1994!
